### PR TITLE
Fix Issue #691 - Flash on Price Change

### PR DESCRIPTION
**Pull request by startailcoon**
Proposal for Issue #691 

Not sure if this is the most efficient model, but it works when I try it.

### CODE REVIEW

**Comment by startailcoon** [comment]
Still to do:
- Consider flash colors
- Consider if more than %-change should flash

File: app/components/Exchange/MyMarkets.jsx [hunk 1]
```diff
@@ -58,12 +59,40 @@ class MarketGroup extends React.Component {
         return (
             !utils.are_equal_shallow(nextState, this.state) ||
             !utils.are_equal_shallow(nextProps.markets, this.props.markets) ||
+            nextState.marketsCache !== this.state.marketsCache ||
```
Comment (svk31) [inline app/components/Exchange/MyMarkets.jsx]: This will always return true, use utils.are_equal_shallow to compare arrays or objects.

**Comment by startailcoon** [comment]
Preview of the function. The delay is most likely due to me debugging and running a lot at the same time. The code adds/removes the desired flash clash and is up to the browser/system.

bitshares-marketflashonchange

**Comment by startailcoon** [comment]
When trying to work with the change on the portfolio dashboard I get the following issue, and I have no idea how to fix it.

image

File: app/components/Account/AccountOverview.jsx [hunk 2]
```diff
@@ -252,6 +278,9 @@ class AccountOverview extends React.Component {
             const canDepositWithdraw = !!this.props.backedCoins.get("OPEN", []).find(a => a.symbol === asset.get("symbol"));
             const canWithdraw = canDepositWithdraw && (hasBalance && balanceObject.get("balance") != 0);
             const canBuy = !!this.props.bridgeCoins.get(symbol);
+            const changeClass = !marketsCache[asset.get("id")] ? "" : marketsCache[asset.get("id")][1] == 0 ? "" : marketsCache[asset.get("id")][1] < 0 ? marketsCache[asset.get("id")][2] ? "pulsate-down" : "change-down" : marketsCache[asset.get("id")][2] ? "pulsate-up" : "change-up";
```
Comment (svk31) [inline app/components/Account/AccountOverview.jsx]: I don't think this line does what you think it does. The syntax is incorrect (too many : parts like here: `? "pulsate-down" : "change-down" : marketsCache[asset.get("id")][2]`) and marketsCache[asset.get("id")][1] is a string so you should parse it before comparing it to 0.

File: app/components/Exchange/MarketRow.jsx [hunk 3]
```diff
@@ -88,110 +99,112 @@ class MarketRow extends React.Component {
...
+                    let amount = stats ? stats.volumeBase : 0;
+                    return (
+                        <td onClick={this._onClick.bind(this, marketID)} className="text-right" key={column.index}>
+                            {utils.format_volume(amount)}
+                        </td>
+                    );
+
+                case "change":
+                    let change = utils.format_number(stats && stats.change ? stats.change : 0, 2);
+                    let changeClass = change === "0.00" ? "" : change > 0 ? flash ? " pulsate-up" : " change-up" : flash ? " pulsate-down" : " change-down";
```
Comment (svk31) [inline app/components/Exchange/MarketRow.jsx]: The syntax is incorrect here as well.

File: app/components/Exchange/MarketRow.jsx [hunk 4]
```diff
@@ -88,110 +99,112 @@ class MarketRow extends React.Component {
...
+                        utils.get_asset_price(price.quote.amount, quote, price.base.amount, base, true);
+
+                    let highPrecisionAssets = ["BTC", "OPEN.BTC", "TRADE.BTC", "GOLD", "SILVER"];
+                    let precision = 6;
+                    if (highPrecisionAssets.indexOf(base.get("symbol")) !== -1) {
+                        precision = 8;
+                    }
+
+                    let change_on_price = utils.format_number(stats && stats.change ? stats.change : 0, 2);
+                    let changeClass_on_price = change_on_price === "0.00" ? "" : change_on_price > 0 ? flash ? " pulsate-up" : " change-up" : flash ? " pulsate-down" : " change-down";
```
Comment (svk31) [inline app/components/Exchange/MarketRow.jsx]: And here

File: app/components/Exchange/MyMarkets.jsx [hunk 5]
```diff
@@ -64,6 +65,35 @@ class MarketGroup extends React.Component {
...
+            marketsCache[id][2] = true;
+        }
+        
+        if(marketsCache[id][2] && timestamp-(2*1000) > marketsCache[id][0]) {
+            //console.log("Market Change: " + id + " expired (" + timestamp + ")");
+            marketsCache[id][2] = false;
+        }
+
+        marketsCache[id][0] = timestamp;
+        marketsCache[id][1] = change;
```
Comment (svk31) [inline app/components/Exchange/MyMarkets.jsx]: I recommend using parseFloat here for the change value.

File: app/components/Exchange/MyMarkets.jsx [hunk 6]
```diff
@@ -64,6 +65,35 @@ class MarketGroup extends React.Component {
...
+        let {marketsCache} = this.state;
+        
+        let timestamp = new Date().getTime();
+        let change = utils.format_number(marketStats.get(id) && marketStats.get(id).change ? marketStats.get(id).change : 0, 2);
+
+        if(!marketsCache[id]) { 
+            marketsCache[id] = [timestamp, "0.00", false];
+        }
+        
+        if(change != marketsCache[id][1] && marketsCache[id][1] != "0.00") {
```
Comment (svk31) [inline app/components/Exchange/MyMarkets.jsx]: Parsing as float below means you can use `!== 0` here. In general you should always use strict equalities.

File: app/components/Utility/MarketChangeComponent.jsx [hunk 7]
```diff
@@ -44,14 +44,22 @@ class MarketChangeComponent extends MarketStats {
...
     getValue() {
         let {marketStats} = this.props;
         return marketStats && marketStats.change ? marketStats.change : 0;
     }
 
     render() {
+        let {flash} = this.state;
         let marketChangeValue = this.getValue();
-        let dayChangeClass = parseFloat(marketChangeValue) === 0 ? "" : parseFloat(marketChangeValue) < 0 ? "change-down" : "change-up";
+        let dayChangeClass = parseFloat(marketChangeValue) === 0 ? "" : parseFloat(marketChangeValue) < 0 ? flash ? "pulsate-down" : "change-down" : flash ? "pulsate-up" : "change-up";
```
Comment (svk31) [inline app/components/Utility/MarketChangeComponent.jsx]: This syntax is also incorrect.

**Comment by startailcoon** [comment]
I will rework this in a new branch and make a new PR once completed.

**Pull request closed without merging**